File: java/amp-rabbit/src/main/java/amp/rabbit/topology/AmqpBaseModel.java
package amp.rabbit.topology;

import java.util.Map;

import com.google.common.base.Objects;
import com.google.common.base.Objects.ToStringHelper;

/**
 * Queues and Exchanges share a number of properties. The shared behavior is
 * captured here.
 * 
 * @author jmccune (Berico Technologies)
 * @author Richard Clayton (Berico Technologies)
 * 
 */
public abstract class AmqpBaseModel {

	protected String name;

	/**
	 * Delete queue/exchange when there are no consumers/queues (respectively)
	 * using the queue/exchange.
	 */
	protected boolean isAutoDelete = false;

	/** indicates that the queue/exchange persists when the server restarts */
	protected boolean isDurable = false;

	/** Indicates if the client should create the queue/exchange if it doesn't exist. */
	protected boolean shouldDeclare = true;

	protected Map<String, Object> arguments;

	public AmqpBaseModel() {
	}

	/**
	 * 
	 * @param name
	 * @param isAutoDelete
	 * @param isDurable
	 * @param shouldDeclare
	 * @param virtualHost
	 * @param arguments
	 */
	public AmqpBaseModel(String name, boolean isAutoDelete, boolean isDurable,
			boolean shouldDeclare, Map<String, Object> arguments) {

		this.name = name;
		this.isAutoDelete = isAutoDelete;
		this.isDurable = isDurable;
		this.shouldDeclare = shouldDeclare;
		this.arguments = arguments;
	}

	// -------------------------------------
	// Standard Getters/Setters
	// -------------------------------------

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public boolean isAutoDelete() {
		return isAutoDelete;
	}

	public void setAutoDelete(boolean isAutoDelete) {
		this.isAutoDelete = isAutoDelete;
	}

	public boolean isDurable() {
		return isDurable;
	}

	public void setDurable(boolean isDurable) {
		this.isDurable = isDurable;
	}

	public boolean shouldDeclare() {
		return shouldDeclare;
	}

	public void setShouldDeclare(boolean shouldDeclare) {
		this.shouldDeclare = shouldDeclare;
	}

	public Map<String, Object> getArguments() {
		return arguments;
	}

	public void setArguments(Map<String, Object> arguments) {
		this.arguments = arguments;
	}

	// -------------------------------------
	// Object Methods
	// -------------------------------------

	@Override
	public int hashCode() {
		//Use Guava version rather than our own hash
		
		// all the arguments (not the instance itself) determine equality/match
		Object argsValue = arguments == null ? null : arguments.hashCode();
		return Objects.hashCode(argsValue,isAutoDelete,
				isDurable,name,shouldDeclare);		
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;		
		if (obj == null || getClass() != obj.getClass())
			return false;
		
		AmqpBaseModel other = (AmqpBaseModel) obj;
		return
			Objects.equal(arguments,other.arguments) &&
			isAutoDelete == other.isAutoDelete &&
			isDurable == other.isDurable &&
			Objects.equal(name, other.name) &&
			shouldDeclare == other.shouldDeclare;
		
	}
	
	protected ToStringHelper getCommonStringInfo() {
		return Objects.toStringHelper(this)
		.add("name",name)		
		.add("autoDelete",isAutoDelete)
		.add("durable",isDurable)
		.add("shouldDeclare",shouldDeclare);
	}
}

File: dotnet/source/amp.rabbit/topology/AmqpBaseModel.cs
using System;
using System.Collections;
using System.Collections.Generic;
using amp.utility;

namespace amp.rabbit.topology
{
    /// <summary>
    /// Queues and Exchanges share a number of properties. The shared behavior is
    /// captured here.
    /// </summary>
    public abstract class AmqpBaseModel
    {
        public string Name { get; private set; }

        /// <summary>
        /// Delete queue/exchange when there are no consumers/queues (respectively)
        /// using the queue/exchange.
        /// </summary>
        public bool IsAutoDelete { get; private set; }

        /// <summary>
        /// Indicates that the queue/exchange persists when the server restarts.
        /// </summary>
        public bool IsDurable { get; private set; }

        /// <summary>
        /// Indicates if the client should create the queue/exchange if it doesn't exist.
        /// </summary>
        public bool ShouldDeclare { get; private set; }
        
        public IDictionary<string, object> Arguments { get; private set; }

        /// <summary>
        /// Returns the sames value as Arguments but as a non-generic Dictionary 
        /// as required for use with the the RabbitMQ .Net client.
        /// </summary>
        public IDictionary ArgumentsAsDictionary
        {
            get
            {
                return Arguments == null
                    ? null
                    : Arguments as IDictionary ?? new Dictionary<string, object>(Arguments);
            }
        } 

        protected AmqpBaseModel(string name, bool isAutoDelete, bool isDurable,
            bool shouldDeclare,
            IDictionary<string, Object> arguments)
        {
            Name = name;
            IsAutoDelete = isAutoDelete;
            IsDurable = isDurable;
            ShouldDeclare = shouldDeclare;
            Arguments = arguments;
        }

        protected bool Equals(AmqpBaseModel other)
        {
            return string.Equals(Name, other.Name) 
                && IsAutoDelete.Equals(other.IsAutoDelete) 
                && IsDurable.Equals(other.IsDurable) 
                && ShouldDeclare.Equals(other.ShouldDeclare)
                && HashUtility.Compare(Arguments, other.Arguments);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((AmqpBaseModel) obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = (Name != null ? Name.GetHashCode() : 0);
                hashCode = (hashCode*397) ^ IsAutoDelete.GetHashCode();
                hashCode = (hashCode*397) ^ IsDurable.GetHashCode();
                hashCode = (hashCode*397) ^ ShouldDeclare.GetHashCode();
                hashCode = (hashCode * 397) ^ (Arguments != null ? HashUtility.GetHashCode(Arguments) : 0);
                return hashCode;
            }
        }
    }
}
